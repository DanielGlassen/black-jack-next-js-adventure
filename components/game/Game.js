import React, { useEffect, useState } from 'react';
import styles from '../../styles/Game.module.css';
import Bet from '../bet/Bet';
import Card from '../card/Card';
import Player from '../player/Player';
import { drawCards } from '../../lib/api';
import { gameResult, handValue, playBlackjack } from '../../lib/gameLogic';
import { Alert } from '../alert/Alert';
import {
	getPlayerBalance,
	updatePlayerBalance,
	listenToAuthChanges
} from '../../config/firebase';
import LoadingSpinner from '../loadingSpinner/LoadingSpinner';

const Game = () => {
	const [playerId, setPlayerId] = useState(null);
	const [playerHand, setPlayerHand] = useState([]);
	const [dealerHand, setDealerHand] = useState([]);
	const [deckId, setDeckId] = useState(null);
	const [gameOver, setGameOver] = useState(false);
	const [balance, setBalance] = useState(1000);
	const [bet, setBet] = useState(0);
	const [gamePhase, setGamePhase] = useState('initial');
	const [lastBet, setLastBet] = useState(null);
	const [showDealerCards, setShowDealerCards] = useState(false);
	const [restartButton, setRestartButton] = useState(false);
	const [loadingBalance, setLoadingBalance] = useState(true);

	useEffect(() => {
		listenToAuthChanges(async (userId) => {
			if (userId) {
				setPlayerId(userId);
				const initialBalance = await getPlayerBalance(userId);
				setBalance(initialBalance);
				setLoadingBalance(false);
			} else {
				setPlayerId(null);
				setBalance(1000);
			}
		});
	}, []);

	const handlePlaceBet = async (betAmount) => {
		try {
			const res = await playBlackjack(playerId, betAmount);
			setPlayerHand(res.playerHand);
			setDealerHand(res.dealerHand);
			setDeckId(res.deckId);
			setBet(betAmount);
			setGamePhase('playing');
		} catch (error) {
			console.error('Error placing bet:', error);
		}
	};

	const handleHit = async () => {
		try {
			const newCards = await drawCards(deckId, 1);
			setPlayerHand((prevHand) => [...prevHand, ...newCards]);
			await checkPlayerHand([...playerHand, ...newCards]);
		} catch (error) {
			console.error('Error handling hit:', error);
		}
	};

	const checkPlayerHand = async (updatedHand) => {
		try {
			let currentPlayerHandValue = await handValue(updatedHand);

			if (currentPlayerHandValue > 21) {
				await handleGameEnd('dealer');
			} else if (currentPlayerHandValue === 21) {
				await handleStand();
			}
		} catch (error) {
			console.error('Error checking player hand:', error);
		}
	};

	const handleStand = async () => {
		try {
			setShowDealerCards(true);
			setGamePhase('stand');
		} catch (error) {
			console.error('Error handling stand:', error);
		}
	};

	useEffect(() => {
		const determineResult = async () => {
			const result = await gameResult(playerHand, dealerHand, deckId);
			await handleGameEnd(result);
		};

		if (gamePhase === 'stand' && dealerHand.length === 2) {
			determineResult();
		}
	}, [dealerHand, gamePhase]);

	const handleGameEnd = async (result, isBj = false) => {
		if (result === null) {
			return;
		}

		setGameOver(true);

		let message = '';
		let messageType = 'success';
		let balanceChange = 0;
		const playerValue = await handValue(playerHand);
		const dealerValue = await handValue(dealerHand);

		if (playerValue > 21) {
			message = 'You lose - busted!';
			balanceChange = -bet;
			messageType = 'error';
		} else if (dealerValue > 21) {
			message = 'You win - dealer busted!';
			balanceChange = bet;
		} else {
			if (isBj === true) {
				message = 'Blackjack!';
				balanceChange = bet * 1.5;
			} else {
				if (result === 'player') {
					message = 'You win!';
					balanceChange = bet;
				} else if (result === 'dealer') {
					message = 'You lose!';
					balanceChange = -bet;
					messageType = 'error';
				} else {
					message = "It's a draw!";
					messageType = 'warning';
				}
			}
		}

		Alert({ message, type: messageType });

		const newBalance = balance + balanceChange;
		setBalance(newBalance);

		try {
			await updatePlayerBalance(playerId, newBalance);
		} catch (error) {
			console.error('Error updating balance in Firebase:', error);
		}

		setRestartButton(true);
		setLastBet(bet);
	};

	const handleRestart = () => {
		setPlayerHand([]);
		setDealerHand([]);
		setGameOver(false);
		setBet(0);
		setDeckId(null);
		setGamePhase('initial');
		setShowDealerCards(false);
	};

	return (
		<>
			{loadingBalance ? (
				<LoadingSpinner />
			) : (
				<>
					{/* Game UI */}
					{balance < 10 && (
						<div className="game-container text-center">
							<div
								className={
									styles.jumbotron +
									' d-flex flex-column align-items-center justify-content-center'
								}
							>
								<h4 className={'display-6'}>
									Your balance is less than $10, you can
									refresh the page to join the new game.
								</h4>
							</div>
						</div>
					)}
					{balance >= 10 && (
						<div className="game-container text-center">
							<div className="dealer">
								{gamePhase !== 'initial' && <h2>Dealer</h2>}
								<div className="dealer-hand">
									{dealerHand.map((card, index) => (
										<Card
											key={index}
											card={card}
											hidden={
												index === 0 && !showDealerCards
											}
										/>
									))}
								</div>
							</div>
							<div className="player mt-5">
								{gamePhase !== 'initial' && <h2>Player</h2>}
								<Player
									playerId={playerId}
									balance={balance}
									bet={bet}
									loading={loadingBalance}
								/>
								{gamePhase === 'playing' ? (
									<p>Current Bet: ${bet}</p>
								) : (
									''
								)}
								{gamePhase === 'playing' && (
									<div className="player-hand">
										{playerHand.map((card, index) => (
											<Card key={index} card={card} />
										))}
									</div>
								)}
								{gamePhase === 'initial' && !gameOver && (
									<Bet
										onPlaceBet={handlePlaceBet}
										lastBet={lastBet}
										balance={balance}
									/>
								)}
							</div>
							{gamePhase === 'playing' && (
								<div className="buttonArea d-flex flex-row align-items-center justify-content-center mt-4">
									{!gameOver && (
										<>
											<button
												className={
													'standButton btn btn-outline-warning px-4 me-3'
												}
												onClick={handleStand}
											>
												Stand
											</button>
											<button
												className={
													'hitButton btn btn-success px-4'
												}
												onClick={handleHit}
											>
												Hit
											</button>
										</>
									)}
								</div>
							)}
							{gameOver && (
								<div className="buttonArea d-flex flex-row align-items-center justify-content-center mt-4">
									<button
										className={
											'restartButton btn btn-outline-primary px-4'
										}
										onClick={handleRestart}
									>
										Restart
									</button>
								</div>
							)}
						</div>
					)}
				</>
			)}
		</>
	);
};

export default Game;
