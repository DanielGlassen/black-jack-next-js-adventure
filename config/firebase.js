import { initializeApp } from 'firebase/app';
import {
	getAuth,
	createUserWithEmailAndPassword,
	signInWithEmailAndPassword,
	onAuthStateChanged,
	signOut
} from 'firebase/auth';
import { getDatabase, ref, set, get } from 'firebase/database';

const firebaseConfig = {
	apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
	authDomain: process.env.NEXT_PUBLIC_AUTH_DOMAIN,
	projectId: process.env.NEXT_PUBLIC_PROJECT_ID,
	storageBucket: process.env.NEXT_PUBLIC_STORAGE_BUCKET,
	messagingSenderId: process.env.NEXT_PUBLIC_MESSAGE_SENDER_ID,
	appId: process.env.NEXT_PUBLIC_APP_ID
  };
  

const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const database = getDatabase(app);

export const getPlayerBalance = async (userId) => {
	const balanceRef = ref(database, `balances/${userId}`);
	try {
		const snapshot = await get(balanceRef);
		if (snapshot.exists()) {
			return snapshot.val();
		} else {
			console.warn(`User not found for userId ${userId}.`);
			return 1000;
		}
	} catch (error) {
		console.error('Error fetching player balance:', error);
		throw error;
	}
};

export const updatePlayerBalance = async (userId, newBalance) => {
	const balanceRef = ref(database, `balances/${userId}`);
	try {
		await set(balanceRef, newBalance);
	} catch (error) {
		console.error(`Error updating balance for user ${userId}:`, error);
		throw error;
	}
};

export const listenToAuthChanges = (callback) => {
	onAuthStateChanged(auth, async (user) => {
		if (user) {
			const userId = user.uid;
			callback(userId);
		} else {
			callback(null);
		}
	});
};

export {
	getAuth,
	createUserWithEmailAndPassword,
	signInWithEmailAndPassword,
	onAuthStateChanged,
	auth,
	signOut,
	database
};
