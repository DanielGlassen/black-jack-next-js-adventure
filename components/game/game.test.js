import React from 'react';
import renderer from 'react-test-renderer';
import Game from './Game';

jest.mock('firebase/app', () => ({
	initializeApp: jest.fn()
}));

jest.mock('firebase/auth', () => ({
	getAuth: jest.fn()
}));

jest.mock('firebase/database', () => ({
	getDatabase: jest.fn()
}));

process.env.NEXT_PUBLIC_Firebase_API_Key = 'test-api-key';
process.env.NEXT_PUBLIC_Auth_Domain = 'test-auth-domain';
process.env.NEXT_PUBLIC_Project_Id = 'test-project-id';
process.env.NEXT_PUBLIC_Storage_Bucket = 'test-storage-bucket';
process.env.NEXT_PUBLIC_Message_Sender_Id = 'test-message-sender-id';
process.env.NEXT_PUBLIC_App_Id = 'test-app-id';

describe('<Game>', () => {
	it('should render component', () => {
		const mockApp = {};
		const mockAuth = {};

		const initializeAppMock = jest.fn(() => mockApp);
		const getAuthMock = jest.fn(() => mockAuth);

		require('firebase/app').initializeApp = initializeAppMock;
		require('firebase/auth').getAuth = getAuthMock;

		const tree = renderer.create(<Game />);
		expect(tree.toJSON()).toMatchSnapshot();
	});
});
