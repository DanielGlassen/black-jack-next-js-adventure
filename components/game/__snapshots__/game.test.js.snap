// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game> renders correctly 1`] = `<MockLoadingSpinner />`;

exports[`<Game> should render component 1`] = `
<div
  className="d-flex justify-content-center align-items-center vh-100"
>
  <div
    className="spinner-border text-primary"
    role="status"
  >
    <span
      className="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;
